albatross/core/traits.h

```diff
@@ -28,36 +28,82 @@ namespace albatross {
  */
 template <class T> struct delay_static_assert : std::false_type {};
 
+/*
+ * In CovarianceFunction we frequently inspect for definitions of
+ * call_impl_ which MUST be defined for const references to objects
+ * (so that repeated covariance matrix evaluations return the same thing
+ *  and so the computations are not repeatedly copying.)
+ * This type conversion utility will turn a type `T` into `const T&`
+ */
+template <class T> struct call_impl_arg_type {
+  typedef
+      typename std::add_lvalue_reference<typename std::add_const<T>::type>::type
+          type;
+};
+
 /*
  * This determines whether or not a class has a method defined for,
  *   `operator() (X x, Y y, Z z, ...)`
  * The result of the inspection gets stored in the member `value`.
  */
 template <typename T, typename... Args> class has_call_operator {
+
   template <typename C,
-            typename = decltype(std::declval<C>()(std::declval<Args>()...))>
-  static std::true_type test(int);
-  template <typename C> static std::false_type test(...);
+            typename = decltype(std::declval<C>()(
+                std::declval<typename call_impl_arg_type<Args>::type>()...))>
+  static std::true_type test(C *);
+  template <typename> static std::false_type test(...);
 
 public:
   static constexpr bool value = decltype(test<T>(0))::value;
 };
 
 /*
  * This determines whether or not a class has a method defined for,
- *   `call_impl_(X x, Y y, Z z, ...)`
+ *   `double call_impl_(const X &x, const Y &y, const Z &z, ...)`
  * The result of the inspection gets stored in the member `value`.
  */
-template <typename T, typename... Args> class has_defined_call_impl {
+template <typename T, typename... Args> class has_valid_call_impl {
+
+  template <typename C>
+  static auto test(C *) -> typename std::is_same<
+      decltype(std::declval<const C>().call_impl_(
+          std::declval<typename call_impl_arg_type<Args>::type>()...)),
+      double>::type;
+  template <typename> static std::false_type test(...);
+
+public:
+  static constexpr bool value = decltype(test<T>(0))::value;
+};
+
+/*
+ * This determines whether or not a class has a method defined for
+ * something close to, but not quite, a valid call_impl_.  For example
+ * if a class has:
+ *   double call_impl_(const X x)
+ * or
+ *   double call_impl_(X &x)
+ * or
+ *   int call_impl_(const X &x)
+ * those are nearly correct but the required `const X &x` in which
+ * case this trait can be used to warn the user.
+ */
+template <typename T, typename... Args> class has_possible_call_impl {
```

Comment:
So this only checks if the `call_impl_()` takes in references as arguments? Will is do anything if there's not return type? Is it an issue if it doesn't?

Reply:
This checks if it takes anything that *could* be a reference (so it'll return true for `call_impl_(X x)` or `call_impl_(X &x)` ...).  I don't think it'll deal with `call_impl_(X &&x)` but I'm ok with that at the moment.

In terms of return type, you're right it doesn't check the return type on purpose.  Since it's used as a way to generate more useful compiler errors I want it to catch some of the more common bugs, things like `int call_impl_(...)` which won't pass the `has_valid_call_impl` test in order to point the user towards the culprit function definition quicker.